src/near_dupe_test.c

```diff
@@ -10,7 +10,9 @@ int main(int argc, char **argv) {
         exit(EXIT_FAILURE);
     }
 
-    if (!libpostal_setup() || !libpostal_setup_language_classifier()) {
+    libpostal_t *instance = libpostal_setup();
+    language_classifier_t *classifier = libpostal_setup_language_classifier();
+    if (instance == NULL || classifier == NULL) {
```

Comment:
This seems like a good point. `language_classifier_t` functions appear to only perform non-mutable operations so storing this object into `libpostal_t` is definitely worth it. Updating the code then!

Reply:
Actually, that doesn't sound *that* good: in the current configuration, you can setup multiple classifiers and also, it "forces" user to call the method in order to get the "language_classifier_t" instance.

Reply:
To extend a bit my last comment: we could maybe go the other way around: wrapping a `libpostal_t` instance into `language_classifier_t`, but then, to be safe, we'd need to add back some globals or shared variables to be sure that `libpostal_t` isn't freed before `language_classifier_t`.

The API looks a bit less nice, I definitely agree with you on this point, but I don't see any good enough solution to make this work without big downsides.

Reply:
Users can still setup multiple classifiers if they have separate instances of libpostal.
In regards to forcing users to call the setup, the library already requires users to call extra setup functions for large modules.

I would try to have only a single extra instance parameter as people upgrading from previous versions of libpostal to an updated one will want the minimum amount of code refactoring.
Being that the project has the aim of being easy to use (shipping with pre-trained model), we should aim to have a minimal impact on people already implementing the library.

Reply:
This is a fair point. Well, I warned about it. :) I'll make the change then but I really don't think this is a good idea.

Also, the bindings have less than 50 functions to bind so it's pretty simple to update. ;)

Reply:
I just can't keep stop thinking about it: the API isn't great. For example: `libpostal_place_languages` is taking two char**. Which doesn't make sense since the first one is suppose to "match" the second, why not an array of `element { label: char*, value: char* }`?

So anyway, the changes remain small, even if we keep the `language_classifier_t` type independant. So why not going to the bottom of this change and prevent all possible data races?

Reply:
> I just can't keep stop thinking about it: the API isn't great. For example: `libpostal_place_languages` is taking two char**. Which doesn't make sense since the first one is suppose to "match" the second, why not an array of `element { label: char*, value: char* }`?

I think a lot of the previous API decisions have been based around the language bindings. I'm going to assume for some language bindings it's easier to have separate labels and values arrays then convert them into a language native datatype.

I'll try to go through the rest of this change request tomorrow and give any feedback on other areas. I think this having an instanced API is a great step forward. And will help the project towards being thread-safe.

Reply:
Ok, we can go back to this discussion once you're done. But I'd really prefer to keep it as is for the previously mentioned reasons.